src/macros.cr

```diff
@@ -83,6 +83,25 @@ macro record(name, *properties)
   end
 end
 
+private record TypeSentinel
+
+class Array(T)
+  macro collect(expr)
+    %arr = [] of typeof(begin
+      %item = TypeSentinel.new
+      {{expr}} do |x|
+        %item = x
+      end
+      raise "Impossible" if %item.is_a? TypeSentinel
```

Comment:
yeah, I mean `raise` without a message would do it. There is no benefit in having that string.